google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessagesProcessor.java

```diff
@@ -257,17 +248,16 @@ public void processReceivedMessages(List<com.google.pubsub.v1.ReceivedMessage> r
     }
     Instant now = new Instant(clock.millis());
     int totalByteCount = 0;
-    final List<AckHandler> ackHandlers = new ArrayList<>(responseMessages.size());
+    final ArrayList<AckHandler> ackHandlers = new ArrayList<>(responseMessages.size());
     for (ReceivedMessage pubsubMessage : responseMessages) {
       int messageSize = pubsubMessage.getMessage().getSerializedSize();
       totalByteCount += messageSize;
       ackHandlers.add(new AckHandler(pubsubMessage.getAckId(), messageSize));
     }
-    ExpirationInfo expiration =
-        new ExpirationInfo(
-            clock, now.plus(messageDeadlineSeconds * 1000), INITIAL_ACK_DEADLINE_EXTENSION_SECONDS);
+    Instant expiration = now.plus(messageDeadlineSeconds * 1000);
     synchronized (outstandingAckHandlers) {
-      addOutstadingAckHandlers(expiration, ackHandlers);
+      outstandingAckHandlers.add(
+          new ExtensionJob(expiration, INITIAL_ACK_DEADLINE_EXTENSION_SECONDS, ackHandlers));
```

Comment:
It just occurred to me that this can cause a race. The AckDeadlineAlarm might access ackHandlers at the same time as this function iterates over it just below. For some reason, moving this synchronized-block just above the try-block below causes a deadlock 100% of the time. I don't fully understand this yet; I'll take another look in the morning.